### PR TITLE
Removes all lag forever

### PR DESCRIPTION
**Pull request by 1fbff5f83b23d39d38b1dfcb4cac8d9b**
Uses the magic of [Lunar Dreamland](https://github.com/goonstation/Lunar-Dreamland) condensed into a small package to ensure BYOND always has time to send map updates to players. It does so by measuring time taken to do that last tick and reserves enough time in the current one. This causes subsystems to process a bit slower but prevents 95% of lagspikes. (statistic pulled out of back orifice)
:cl: McDonald072
rscdel: Lag
/:cl:

If you want to port this to another codebase, make sure that you place _maptick_api somewhere it will be compiled before your master controller, and that you call `maptick_initialize()` and `maptick_shutdown()` in `world/New()` and `world/Reboot()` respectively.

### CODE REVIEW

**Comment by gilvian** [comment]
we topped every PR, thats it. close the codebase

**Comment by Bibby0110** [comment]
Stupid idea, but what about adding the .dll source code?

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
Source for both DLLs available here: https://github.com/goonstation/Lunar-Dreamland

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
Travis failed because it couldn't load the library, but it compiles and I tested it locally

**Comment by ynot01** [comment]
we did it boys we stopped the ceo of lag

**Comment by Altoids1** [comment]
Just for those who aren't in the know about this, this Lunar Dreamland stuff genuinely has the potential to completely resolve the lag situation and massively improve the performance of the game. If Mac can get this to work, long-term we can actually reduce lag by 90% or more.

So, yeah, Literally The Best PR. Love it.

**Comment by oklews** [comment]
We’re in the endgame now

**Comment by Couls** [comment]
image
Running your branch
image

File: code/game/world.dm [hunk 1]
```diff
@@ -48,6 +48,9 @@ GLOBAL_VAR(restart_counter)
 
 	GLOB.timezoneOffset = text2num(time2text(0,"hh")) * 36000
 
+	world.log << call("hookerino.dll","BHOOK_Init")()
```
Comment (Couls) [inline code/game/world.dm]: shouldn't this be uppercase H?

**Comment by yacabo** [comment]
I believe in you, make this work, spite Travis

**Comment by ThatLing** [comment]
You should really clean these lua files up

File: code/game/world.dm (hunk 1 above)
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [reply code/game/world.dm]: on windows the case in paths doesn't matter

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
@Couls I'm aware of that issue but it works on my machine™ so I'm having a hard time even starting to debug it, sadly

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
CHANGES!!

Threw out all that Lua crap and packaged everything into a single DLL. Now it should work better and faster ~~and stronger~~. It also should be very easy to port to all codebases that rely on measuring tick usage. Source code coming soon.

**Comment by Bibby0110** [comment]
Inb4 Yogstation gets banned for EULA violations

**Comment by Couls** [comment]
Can confirm it works with the new dlls, tested on paradise codebase

**Comment by Bibby0110** [comment]
tHIS IS VERY NICE

**Comment by AsV9** [comment]
I'd like to look at those .dlls aswell.

**Comment by Readystorm546** [comment]
Travis is not happy

**Comment by optimumtact** [comment]
if I were you, I'd hard reserve at least 20% of a tick for your game code.

SSInput or SStimer missing ticks will cause all sorts of really noticeable issues for players.

Actually better in those cases to go over a tick

**Comment by actioninja** [comment]
https://github.com/BeeStation/BeeStation-Hornet/pull/412#issuecomment-530555134
relevant comment from beestation devs after testing

**Comment by ghost** [comment]
Would this work on Linux, if it was compiled for it?

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
I'd assume if you made the necessary modifications then yes, but I don't know how it's implemented on linux

**Comment by ghost** [comment]
I THINK lunar dreamland has Linux support, judging from all the 
```
#ifdef _WIN32
#else
#endif
```

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
lunar itself does, but the library I wrote here does not

**Comment by ghost** [comment]
oh.

Uhhh where's the source for said lib

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
proprietary

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
https://github.com/goonstation/Lunar-Dreamland/tree/master/maptick

You can see the source for the library here for now

**Comment by alexkar598** [comment]
Stale, said he will be coming back to it. Closing PR until then.

**Pull request closed without merging**